File: apps/android/collaboration/src/main/kotlin/diy/atelier/collaboration/CollaborationEngine.kt
package diy.atelier.collaboration

data class CollaborationProtocolVersion(
    val major: UShort,
    val minor: UShort,
) {
    companion object {
        val Foundation = CollaborationProtocolVersion(major = 0u, minor = 1u)
    }
}

data class CollaborationOperation(
    val documentId: String,
    val authorDid: String,
    val sequence: ULong,
    val payload: ByteArray,
    val version: CollaborationProtocolVersion = CollaborationProtocolVersion.Foundation,
)

sealed interface CollaborationAvailability {
    data object NotConfigured : CollaborationAvailability
    data object Available : CollaborationAvailability
    data class Unavailable(val reason: String) : CollaborationAvailability
}

interface CollaborationEngine {
    suspend fun availability(): CollaborationAvailability
    suspend fun apply(operation: CollaborationOperation)
}

File: apps/apple/Packages/AtelierShared/Sources/AtelierCollaboration/CollaborationEngine.swift
import AtelierCore
import Foundation

public struct CollaborationProtocolVersion: Codable, Equatable, Sendable {
    public let major: UInt16
    public let minor: UInt16

    public init(major: UInt16, minor: UInt16) {
        self.major = major
        self.minor = minor
    }

    public static let foundation = CollaborationProtocolVersion(major: 0, minor: 1)
}

public struct CollaborationOperation: Codable, Equatable, Sendable {
    public let documentID: UUID
    public let authorDID: String
    public let sequence: UInt64
    public let payload: Data
    public let version: CollaborationProtocolVersion

    public init(
        documentID: UUID,
        authorDID: String,
        sequence: UInt64,
        payload: Data,
        version: CollaborationProtocolVersion = .foundation
    ) {
        self.documentID = documentID
        self.authorDID = authorDID
        self.sequence = sequence
        self.payload = payload
        self.version = version
    }
}

public enum CollaborationAvailability: Equatable, Sendable {
    case notConfigured
    case available
    case unavailable(reason: String)
}

public protocol CollaborationEngine: Sendable {
    func availability() async -> CollaborationAvailability
    func apply(_ operation: CollaborationOperation) async throws
}
